src/java.base/share/classes/java/lang/StringConcatHelper.java

```diff
@@ -440,4 +571,181 @@ static MethodHandle lookupStatic(String name, MethodType methodType) {
+        }
+        return index;
+    }
+
+    /**
+     * Prepends constant and the stringly representation of value into buffer,
+     * given the coder and final index. Index is measured in chars, not in bytes!
+     *
+     * @param index     final char index in the buffer
+     * @param coder     coder of the buffer
+     * @param buf        buffer to append to
+     * @param value      int value to encode
+     * @param prefix     a constant to prepend before value
+     * @return           updated index
+     */
+    static int prepend(int index, byte coder, byte[] buf, int value, String prefix) {
+        if (coder == String.LATIN1) {
+            index = StringLatin1.getChars(value, index, buf);
+            index -= prefix.length();
+            prefix.getBytes(buf, index, String.LATIN1);
+        } else {
+            index = StringUTF16.getChars(value, index, buf);
+            index -= prefix.length();
+            prefix.getBytes(buf, index, String.UTF16);
+        }
+        return index;
+    }
+
+    /**
+     * Prepends constant and the stringly representation of value into buffer,
+     * given the coder and final index. Index is measured in chars, not in bytes!
+     *
+     * @param index     final char index in the buffer
+     * @param coder     coder of the buffer
+     * @param buf        buffer to append to
+     * @param value      long value to encode
+     * @param prefix     a constant to prepend before value
+     * @return           updated index
+     */
+    static int prepend(int index, byte coder, byte[] buf, long value, String prefix) {
+        if (coder == String.LATIN1) {
+            index = StringLatin1.getChars(value, index, buf);
+            index -= prefix.length();
+            prefix.getBytes(buf, index, String.LATIN1);
+        } else {
+            index = StringUTF16.getChars(value, index, buf);
+            index -= prefix.length();
+            prefix.getBytes(buf, index, String.UTF16);
+        }
+        return index;
+    }
+
+    /**
+     * Prepends constant and the stringly representation of value into buffer,
+     * given the coder and final index. Index is measured in chars, not in bytes!
+     *
+     * @param index     final char index in the buffer
+     * @param coder     coder of the buffer
+     * @param buf        buffer to append to
+     * @param value      boolean value to encode
+     * @param prefix     a constant to prepend before value
+     * @return           updated index
+     */
+    static int prepend(int index, byte coder, byte[] buf, String value, String prefix) {
+        index -= value.length();
+        if (coder == String.LATIN1) {
+            value.getBytes(buf, index, String.LATIN1);
+            index -= prefix.length();
+            prefix.getBytes(buf, index, String.LATIN1);
+        } else {
+            value.getBytes(buf, index, String.UTF16);
+            index -= prefix.length();
+            prefix.getBytes(buf, index, String.UTF16);
+        }
```

Comment:
Similar to above, can if/else be collapsed and inline `coder`?

```suggestion
        value.getBytes(buf, index, coder);
        index -= prefix.length();
        prefix.getBytes(buf, index, coder);
```